File: CSharp/DifferenceBetweenFibonacciAndFactorial/DifferenceBetweenFibonacciAndFactorial/FibonacciEngine/FibonacciEngine.cs
namespace DifferenceBetweenFibonacciAndFactorial.FibonacciEngine
{
    public class FibonacciEngine : IFibonacciEngine
    {
        public int GetFibonacciNumberRecursive(int indexOfNumber)
        {
            if (IsEqualWithOneOrTwo(indexOfNumber))
                return 1;

            return GetFibonacciNumberRecursive(indexOfNumber - 1) + GetFibonacciNumberRecursive(indexOfNumber - 2);
        }

        private static bool IsEqualWithOneOrTwo(int indexOfNumber)
        {
            return indexOfNumber == 1 || indexOfNumber == 2;
        }

        public int GetFibonacciNumber(int indexOfNumber)
        {
            var firstTerm = 0;
            var secondTerm = 1;

            if (indexOfNumber == 0) return 1;

            secondTerm = ProcessSwapOfElements(indexOfNumber, firstTerm, secondTerm);

            return secondTerm;
        }

        private static int ProcessSwapOfElements(int indexOfNumber, int firstTerm, int secondTerm)
        {
            for (int it = 2; it <= indexOfNumber; it++)
                firstTerm = SwapFibonacciElements(firstTerm, ref secondTerm);
            
            return secondTerm;
        }

        private static int SwapFibonacciElements(int firstTerm, ref int secondTerm)
        {
            var thirdTerm = firstTerm + secondTerm;
            firstTerm = secondTerm;
            secondTerm = thirdTerm;
            
            return firstTerm;
        }
    }
}

File: Python/difference_between_fibonacci_and_factorial/fibonacci_engine.py
class IFibonacciEngine:
    def get_fibonacci_number(self, index: int):
        pass


class FibonacciEngine(IFibonacciEngine):

    @staticmethod
    def __is_value_one_or_zero(index: int):

        return index == 1 or index == 2

    def get_fibonacci_number(self, index: int):

        if self.__is_value_one_or_zero(index):
            return 1

        return self.get_fibonacci_number(index - 1) + self.get_fibonacci_number(index - 2)

